### ISSUE TITLE
Pass object and prototype as parameter

### ISSUE THREAD

**Issue by Adavo** [vkiryukhin/ng-vkthread#1]
Hello,

I use AngularJS 1 and Typescript and I have a list of 200 000 objects to process which freeze the DOM for around 3 seconds on mobile. 

I saw your work (which look amazing) after fighting so long with web worker.
My main difficluty is to pass a typescript class to the worker.

Here is my code:

```
var vkthread = this.vkThread();
console.log(Word.prototype);
vkthread.exec({
      fn: (Word, data) => {
              console.log(Word.prototype);
              let externalDicServerObject = JSON.parse(data);
              let result: Word[] = [];
              for (let wordData of externalDicServerObject) {
                     result.push(Word.initializeFromServerObject(wordData));
              }
              return result;
        }, args: [Word, data]
 }).then(
      (data: any) => {
             console.log("yes");
       },
       (err) => {
            console.log(err);
      }
);
```

This code provoke this issue:

> Error: Worker error: Uncaught TypeError: Word.initializeFromServerObject is not a function
    at Worker.worker.onerror (vkthread.min.js:15)

But what is interesting is the console.log of the prototype of Word.
The first console display:

> Object {fixPinyin: function, initializeFromServerObject: function, toString: function, getTranslation: function, isRecognizedCharacter: function…}

but the second:
> Object {constructor: function}

So it seems that the argument sent through Vkthread is not fully restored on the other side...
If I send the prototype directly, then I can access my function but my object attribute is not initialized so I got some issue.

I don't have any idea how I can fix this issue.
Thanks for your help.

**Comment by vkiryukhin** [comment]
Hi Adavo,
As I don't know content of the Word object, I can't say anything useful
now. As far as I understood this was an object with multiple member
functions which you wanted to use in worker, right? I need to refresh my
memory to remember if I tested such kind of scenario. Maybe I'll spend some
time on that on weekend.

best,
--Vadim

**Comment by Adavo** [comment]
Yup, indeed, Word is a class in typescript which in pure javascript is a function with prototype function.
If you success to reproduce (and fix) this issue, thanks for letting me know.
On my side I have tried to serialize the whole function but still have problem ☹

**Comment by vkiryukhin** [comment]
Hi Adavo,
try to configure your task this way: group those functions in object and pass this object as a context. Call those function with `this` prefix. Please try it with plain javascript first: I've never tested this plugin with typescript.

Example:

```
var obj = {
    sum: function(n,m){
        return n+m;
    },
    mul: function(x,y){
        return x*y;
    }
}

function foo(a,b){
    return [this.sum(a,b), this.mul(a,b)];
}

var param = {
    fn: foo,      // <-- function to execute
    args: [3, 2], // <-- arguments for this function
    context: obj
};

vkthread.exec(param).then(
    function (data) {
       console.log(data);  // <-- thread returns [5, 6]
    },
    function(err) {
        alert(err);  // <-- thread returns error message
    }
);
```

hope this helps,

--Vadim
